### PR TITLE
Allow organizers to create and delete lectures for an event

### PR DESCRIPTION
**Pull request by MarioRodrigues10**

### CODE REVIEW

**Review by ruioliveira02 — COMMENTED** [review]
Make sure the CI passes:

```npm run lint```

File: apps/app/components/AppMenu/index.tsx [hunk 1]
```diff
@@ -137,6 +137,8 @@ function AppMenu({ hidePrimaryMenu, collapsed }: any) {
                 <SubMenu icon={<CalendarOutlined />} title="Eventos">
                   <Item key="/events">Listar eventos</Item>
                   <Item key="/admin/event">Criar evento</Item>
+                  <Item key="/admin/lectures">Listar lectures</Item>
```
Comment (RuiL1904) [inline apps/app/components/AppMenu/index.tsx:140]:
```suggestion
                  <Item key="/admin/lectures">Listar sessões</Item>
```

File: apps/app/components/AppMenu/index.tsx [hunk 2]
```diff
@@ -137,6 +137,8 @@ function AppMenu({ hidePrimaryMenu, collapsed }: any) {
...
+                  <Item key="/admin/lectures/new">Criar Lecture</Item>
```
Comment (RuiL1904) [inline apps/app/components/AppMenu/index.tsx:141]:
```suggestion
                  <Item key="/admin/lectures/new">Criar sessão</Item>
```

File: apps/app/components/Event/index.js [hunk 3]
```diff
@@ -122,9 +122,14 @@ const Event = ({
...
             {role === EUser.Organizer ? (
-              <Link href={`/admin/event/${event.id}`}>
-                <Button type="primary">Info</Button>
-              </Link>
+              <>
+                <Link href={`/admin/event/${event.id}`}>
+                  <Button type="primary">Info</Button>
+                </Link>
+                <Link href={`/admin/lecture/${event.id}`}>
+                  <Button type="primary">Criar Lecture</Button>
```
Comment (RuiL1904) [inline apps/app/components/Event/index.js:130]: This button its not properly working.

File: packages/bokkenjs/lib/events/events.ts [hunk 4]
```diff
@@ -37,3 +37,9 @@ export async function createEvent(values: any) {
 
   return response.data;
 }
+
+export async function list_events() {
```
Comment (RuiL1904) [inline packages/bokkenjs/lib/events/events.ts:41]:
```suggestion
export async function listEvents() {
```

File: packages/bokkenjs/lib/users/mentors.ts [hunk 5]
```diff
@@ -5,3 +5,9 @@ export async function getMentor(id: string) {
 
   return response.data;
 }
+
+export async function list_mentors() {
```
Comment (RuiL1904) [inline packages/bokkenjs/lib/users/mentors.ts:9]:
```suggestion
export async function listMentors() {
```

File: apps/app/pages/admin/lectures/index.tsx [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+          style={{ width: 200 }}
+          onChange={setSelectedEvent}
+          value={selectedEvent == "" ? undefined : selectedEvent}
+        >
+          {events.map((event) => (
+            <Select.Option value={event.id}>{event.title}</Select.Option>
+          ))}
+        </Select>
+      </Row>
+      {selectedLectures.map((lecture) => (
```
Comment (RuiL1904) [inline apps/app/pages/admin/lectures/index.tsx:86]: You can use `Card` component from ant-design in order to achieve this.

File: apps/app/components/AppMenu/index.tsx [hunk 7]
```diff
@@ -137,6 +137,8 @@ function AppMenu({ hidePrimaryMenu, collapsed }: any) {
                 <SubMenu icon={<CalendarOutlined />} title="Eventos">
                   <Item key="/events">Listar eventos</Item>
                   <Item key="/admin/event">Criar evento</Item>
+                  <Item key="/admin/lectures">Listar sessões</Item>
+                  <Item key="/admin/lectures/new">Criar sessão</Item>
```
Comment (RuiL1904) [inline apps/app/components/AppMenu/index.tsx:141]: I think this would be prettier on a `SubMenu` called "Sessões". Following what's on a mentor a view, for example.

File: apps/app/components/Event/index.js [hunk 8]
```diff
@@ -122,9 +122,14 @@ const Event = ({
...
             {role === EUser.Organizer ? (
-              <Link href={`/admin/event/${event.id}`}>
-                <Button type="primary">Info</Button>
-              </Link>
+              <>
+                <Link href={`/admin/event/${event.id}`}>
+                  <Button type="primary">Info</Button>
+                </Link>
+                <Link href={`/admin/lectures/new`}>
+                  <Button type="primary">Criar Lecture</Button>
```
Comment (RuiL1904) [inline apps/app/components/Event/index.js:130]: This button only makes sense if the new lecture page can receive a prop called, for instance, `event_id`. Otherwise, it's just a general redirect and you can remove it.

File: apps/app/components/LectureForm/index.js [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+    }
+  };
+
+  return (
+    <>
+      <Row justify="space-between">
+        <Title level={2}>
+          {id && ninjas[0]
+            ? ninjas[0].first_name + " " + ninjas[0].last_name
+            : "Nova Sessão"}
```
Comment (RuiL1904) [inline apps/app/components/LectureForm/index.js:86]:
```suggestion
          Nova Sessão"
```

File: apps/app/components/LectureForm/index.js [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+            onFinish={onFinish}
+          >
+            <Form.Item
+              label="Evento"
+              name="lecture[event_id]"
+              rules={[{ required: true }]}
+            >
+              <Select
+                placeholder="Escolha um evento"
+                style={{ width: "50%" }}
```
Comment (RuiL1904) [inline apps/app/components/LectureForm/index.js:123]:
```suggestion
                style={{ width: "50%" }}
```

Can you put a higher width? Same for the other items with such width.

File: apps/app/components/LectureForm/index.js [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+                placeholder="Escolha pelo menos um Ninja"
+                style={{ width: "50%" }}
+                onChange={setSelectedNinja}
+                value={
+                  Object.keys(selectedNinja).length === 0
+                    ? undefined
+                    : selectedNinja
+                }
+              >
+                {ninjas.map((ninja) => (
```
Comment (RuiL1904) [inline apps/app/components/LectureForm/index.js:156]: Here, should only appear ninjas that doesn't already have a lecture for the current selected event. Otherwise, if I select a ninja and a mentor that already have a lecture I will receive an error from the API saying "já existe no sistema". Bearing that in mind, add the same constraint to the mentors listing.

File: apps/app/pages/admin/lectures/index.tsx [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+            type="primary"
+            size="large"
+            icon={<PlusOutlined />}
+          />
+        </Link>
+      </Row>
+      <Row justify="space-around" gutter={[10, 10]}>
+        <Select
+          defaultValue="Selecione um Evento"
+          style={{ width: 200 }}
```
Comment (RuiL1904) [inline apps/app/pages/admin/lectures/index.tsx:100]: Higher width here too.

File: apps/app/pages/admin/lectures/index.tsx [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+              <Descriptions.Item
+                labelStyle={labelStyle}
+                label={
+                  <span>
+                    <EnvironmentOutlined /> Localização
+                  </span>
+                }
+                span={1}
+              >
+                Departamento de Informática, Edifício 7, Universidade do Minho
```
Comment (RuiL1904) [inline apps/app/pages/admin/lectures/index.tsx:154]: Here should be the current event location (`lecture.event.location`). Not an hard coded one.

File: apps/app/components/AppMenu/index.tsx [hunk 14]
```diff
@@ -137,6 +138,12 @@ function AppMenu({ hidePrimaryMenu, collapsed }: any) {
...
+                </SubMenu>
+                <SubMenu icon={<UsergroupAddOutlined />} title="Sessões">
+                  <Item key="/admin/lectures">Listar sessões</Item>
+                  <Item key="/admin/lectures/new">Criar sessão</Item>
```
Comment (RuiL1904) [inline apps/app/components/AppMenu/index.tsx:146]: I don't think it makes sense to have this buttons (list and create lecture) on both menus

File: apps/app/components/Event/index.js [hunk 15]
```diff
@@ -122,9 +122,14 @@ const Event = ({
...
             {role === EUser.Organizer ? (
-              <Link href={`/admin/event/${event.id}`}>
-                <Button type="primary">Info</Button>
-              </Link>
+              <>
+                <Link href={`/admin/event/${event.id}`}>
+                  <Button type="primary">Info</Button>
+                </Link>
+                <Link href={`/admin/lectures/${event.id}`}>
+                  <Button type="primary">Criar Lecture</Button>
```
Comment (RuiL1904) [inline apps/app/components/Event/index.js:130]:
```suggestion
                  <Button type="primary">Criar sessão</Button>
```

File: apps/app/components/LectureForm/index.js [hunk 16]
```diff
@@ -0,0 +1,218 @@
...
+
+  const onFinish = (values) => {
+    if (Object.keys(selectedEvent).length != 0) {
+      api
+        .createLecture(values)
+        .then(() => {
+          router.push("/admin/lectures");
+        })
+        .catch((error) => {
+          notifyError("Ocorreu um erro", "Não foi possível criar uma lecture");
```
Comment (RuiL1904) [inline apps/app/components/LectureForm/index.js:74]:
```suggestion
          notifyError("Ocorreu um erro", "Não foi possível criar a sessão");
```

File: apps/app/components/LectureForm/index.js [hunk 17]
```diff
@@ -0,0 +1,218 @@
...
+        });
+    } else {
+      api
+        .createLecture(values)
+        .then((response) => {
+          router.push("/admin/lectures");
+        })
+        .catch((error) => {
+          notifyError("Ocorreu um erro", "Não foi possível criar uma lecture");
```
Comment (RuiL1904) [inline apps/app/components/LectureForm/index.js:83]:
```suggestion
          notifyError("Ocorreu um erro", "Não foi possível criar a sessão");
```

File: apps/app/components/LectureForm/index.js [hunk 18]
```diff
@@ -0,0 +1,218 @@
...
+        });
+    }
+  };
+
+  return (
+    <>
+      <Row justify="space-between">
+        <Title level={2}>Nova Sessão</Title>
```
Comment (RuiL1904) [inline apps/app/components/LectureForm/index.js:91]:
```suggestion
        <Title level={2}>Nova sessão</Title>
```

File: apps/app/pages/admin/event.tsx [hunk 19]
```diff
@@ -34,21 +35,14 @@ function CreateEvent() {
     api
       .getLocations()
       .then((response) => setLocations(response.data))
-      .catch((error) => {
-        notifyError(
-          "Ocorreu um erro",
-          "Não foi possível obter as localizações"
-        );
-      });
+      .catch((error: any) => notification["error"](error.data?.errors));
```
Comment (RuiL1904) [inline apps/app/pages/admin/event.tsx:38]: This should be on a `notifyError`, same for the below changes.

File: packages/ui/components/Auth/AuthProvider.tsx [hunk 20]
```diff
@@ -12,6 +12,7 @@ export function AuthProvider({ children }: PropsWithChildren<Props>) {
   const [errors, setErrors] = useState<any | IErrors>();
   const [isLoading, setLoading] = useState<boolean>(false);
   const [isFirstLoading, setFirstLoading] = useState<boolean>(true);
+  console.log(isFirstLoading);
```
Comment (RuiL1904) [inline packages/ui/components/Auth/AuthProvider.tsx:15]:
```suggestion
  console.log(isFirstLoading);
```

Remove this please 🙏🏼

**Review by RuiL1904 — CHANGES_REQUESTED** [review]
There's, probably, a `useEffect` without a dependency. Bokken is being spammed by requests. Fix it, please 🙏🏼

File: apps/app/components/LectureForm/index.js [hunk 21]
```diff
@@ -0,0 +1,234 @@
...
+    });
+    setFilteredNinjas(filtered);
+  }, [ninjas, selectedEvent, lectures]);
+
+  const onFinish = (values) => {
+    if (Object.keys(selectedEvent).length != 0) {
+      api
+        .createLecture(values)
+        .then(() => {
+          notifyInfo("Sessão criada com sucesso.");
```
Comment (RuiL1904) [inline apps/app/components/LectureForm/index.js:94]:
```suggestion
          notifyInfo("Sessão criada com sucesso");
```

File: apps/app/pages/admin/event.tsx [hunk 22]
```diff
@@ -37,7 +38,7 @@ function CreateEvent() {
       .catch((error) => {
         notifyError(
           "Ocorreu um erro",
-          "Não foi possível obter as localizações"
+          "Não foi possível obter as localizações."
```
Comment (RuiL1904) [inline apps/app/pages/admin/event.tsx:41]: I think this should belong to another PR. We could discuss this consistency thought.

File: apps/app/components/LectureForm/index.js [hunk 23]
```diff
@@ -0,0 +1,234 @@
...
+                  value={
+                    Object.keys(selectedEvent).length === 0
+                      ? undefined
+                      : selectedEvent
+                  }
+                >
+                  {events.map((event) => (
+                    <Select.Option key={event.id} value={event.id}>
+                      <div>
+                        {`${event.title} - ${event.start_time} - ${event.end_time}`}
```
Comment (RuiL1904) [inline apps/app/components/LectureForm/index.js:155]: Format the displayed date 🙏🏼

**Review by RuiL1904 — APPROVED** [review]
What a great job 🚀
